---
title: "class09"
author: "Pham Vo"
date: "2/17/2022"
output:
  pdf_document: default
  html_document: default
---

```{r}
pdb <- "Data Export Summary.csv"
pdb.df <- read.csv(pdb)
pdb.df
```
> Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
solved.Xray <- sum(pdb.df$X.ray)
solved.EM <- sum(pdb.df$EM)
solved.total <- sum(pdb.df[,2:8])

percent.solved.Xray.EM <- ((solved.Xray + solved.EM)/solved.total)*100

percent.solved.Xray.EM
```

> Q2: What proportion of structures in the PDB are protein?

```{r}
mol.type.protein <- pdb.df[1,8]
mol.type.total <- sum(pdb.df[1:6,8])

prop.mol.type.protein <- (mol.type.protein/mol.type.total)*100
prop.mol.type.protein
```

> Q3: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

1868

#Introductoion to Bio3D


```{r}
#Load Bio3D package to R
library(bio3d)

#Reading PDB file data into R

pdb <- read.pdb("1hsg")
pdb



```

> Q7: How many amino acid residues are there in this pdb object? 198

> Q8: Name one of the two non-protein residues? HOH

> Q9: How many protein chains are in this structure? 2

```{r}
#To find the attributes of any such object:

attributes(pdb)
```

```{r}
# To access these individual attributes we use the dollar-attribute name convention that is common with R list objects. For example, to access the atom attribute or component use

head(pdb$atom)
```

# Comparative structure analysis of Adenylate Kinase

```{r}
# we analyze all currently available Adk structures in the PDB to reveal detailed features and mechanistic principles of these essential shape changing transitions.

# In terms of protein structures PCA can be used to capture major structural variations within a set of structures (a.k.a. structure ensemble). This can make interpreting major conformational states (such as ‘active’ and ‘inactive’ or ‘ligand bound’ and ‘un-bound’ states) and structural mechanisms for activation or regulation more clear.
```

> Q10. Which of the packages above is found only on BioConductor and not CRAN? msa

> Q11. Which of the above packages is not found on BioConductor or CRAN? Grantlab/bio3d-view

> Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket? TRUE


```{r}
# Search and retrieve ADK structures
## fetch the query sequence for chain A of the PDB ID 1AKE
library(bio3d)
aa <- get.seq("1ake_A")
aa
```

> Q13. How many amino acids are in this sequence, i.e. how long is this sequence? 214

```{r}
# use this sequence as a query to BLAST search the PDB to find similar sequences and structures

# Blast or hmmer search
b <- blast.pdb(aa)

# Plot a summary of search results (adjusting the cutoff argument (to plot.blast()) will result in a decrease or increase of hits)
hits <- plot(b)
```

```{r}
# List out some 'top hits'
head(hits$pdb.id)
```

```{r}
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

# Align and superpose structures

```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE)

# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
#plot(pdbs, labels=ids)
```

```{r}
#Viewing our superposed structures
#library(bio3d.view)
#library(rgl)

#view.pdbs(pdbs)
```

```{r}
#Annotate collected PDB structures
#ids <- basename.pdb(pdbs$id)
#anno <- pdb.annotate(as.vector(ids))
#unique(anno$source)

# View all available annotation data
#anno
```

# Principal component analysis

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```

# Norrmal mode analysis

```{r}
# NMA of all structures
modes <- nma(pdbs)

plot(modes, pdbs, col=grps.rd)
```

> Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

Black and colored lines are different. They are different the most in 2 regions: aa 25-75 and aa 125-175.

These could be nucleotide-binding regions, which is essential to be flexible. 






